test/unit/node/heart.test.ts

```diff
@@ -0,0 +1,98 @@
+import { logger } from "@coder/logger"
+import { readFile, writeFile } from "fs/promises"
+import { Heart, heartbeatTimer } from "../../../src/node/heart"
+import { clean, mockLogger, tmpdir } from "../../utils/helpers"
+
+function mockIsActive(resolveTo: boolean): () => Promise<boolean> {
+  return () => new Promise((resolve, reject) => setTimeout(() => resolve(resolveTo), 100))
+}
+
+describe("Heart", () => {
+  const testName = "heartTests"
+  let testDir = ""
+  let heart: Heart
+
+  beforeAll(async () => {
+    mockLogger()
+    await clean(testName)
+    testDir = await tmpdir(testName)
+  })
+  beforeEach(() => {
+    heart = new Heart(`${testDir}/shutdown.txt`, mockIsActive(true))
+  })
+  afterAll(() => {
+    jest.restoreAllMocks()
+  })
+  afterEach(() => {
+    jest.resetAllMocks()
+    if (heart) {
+      heart.dispose()
+    }
+  })
+  it("should write to a file when given a valid file path", async () => {
+    // Set up heartbeat file with contents
+    const text = "test"
+    const pathToFile = `${testDir}/file.txt`
+    await writeFile(pathToFile, text)
+    const fileContents = await readFile(pathToFile, { encoding: "utf8" })
+    expect(fileContents).toBe(text)
+
+    heart = new Heart(pathToFile, mockIsActive(true))
+    heart.beat()
+    // Check that the heart wrote to the heartbeatFilePath and overwrote our text
+    const fileContentsAfterBeat = await readFile(pathToFile, { encoding: "utf8" })
+    expect(fileContentsAfterBeat).not.toBe(text)
+  })
+  it("should log a warning when given an invalid file path", async () => {
+    heart = new Heart(`fakeDir/fake.txt`, mockIsActive(false))
+    heart.beat()
+    // HACK@jsjoeio - beat has some async logic but is not an async method
+    // Therefore, we have to create an artificial wait in order to make sure
+    // all async code has completed before asserting
```

Comment:
This looks good to me for now.  Maybe long-term we can make `beat` `async`, at least up to the file write.

Reply:
Yes! I briefly looked into it for this PR but since it happens in a request, I wasn't sure if changing it [here](https://github.com/coder/code-server/blob/a1af9e2a5646df1a587f73fbe1455ca28824aeee/src/node/routes/index.ts#L55-L60) to make the handler `async` would causes so I didn't.

Reply:
Ah yeah I think the move would be to only `await` in tests.  For other uses we can leave them as-is (since we do not want to delay the request to write to the heartbeat file).

Reply:
Ah gotcha! I did use `await` when I was first writing the tests but TS said it would have no effect lol but I guess that's because it couldn't infer the callback logic? I'm not sure lol

Reply:
Ahh yup we would need to make `beat` `async` first and then add the `await` in the tests.

Reply:
But I think the current workaround is chill